File: frontend/types/hugContract.ts
/**
 * HUG 사고 전 계약관리 API 응답 타입.
 * backend/app/services/hug_contract_service.py · schemas/hug_contract.py 실응답(260723) 기준.
 */

import type {
  DdayCheckpoint,
  PredictionStatus,
  PreventionStatus,
  PreventiveActionStatus,
} from "@/lib/hug-labels";

export interface AccidentPredictionFactor {
  feature: string;
  label: string;
  value: string | number | null;
  importance: number;
}

/** 계약별 사고위험 산정 결과 — 목록·상세·이력 공용 뷰. */
export interface AccidentPrediction {
  prediction_id: string;
  pu_risk_score: number | null;
  risk_percentile: number | null;
  accident_probability: number | null;
  prediction_status: PredictionStatus;
  failure_reason: string[];
  model_version: string;
  top_factors: AccidentPredictionFactor[];
  data_completeness: number;
  predicted_at: string | null;
  valid_until: string | null;
}

export interface RuleRisk {
  risk_assessment_id: string;
  risk_score: number | null;
  risk_grade: string | null;
  risk_factors: { factor?: string; description?: string; severity?: string; [key: string]: unknown }[];
  data_completeness: number | null;
  created_at: string | null;
}

export interface PreventionCaseSummary {
  prevention_case_id: string;
  status: PreventionStatus;
  triggers: { trigger?: string; reason?: string; [key: string]: unknown }[];
  owner_user_id: string | null;
  owner_center: string | null;
  next_action: string | null;
  due_at: string | null;
}

export interface EvidenceBundleItem {
  item_key: string;
  label: string;
  evidence_type: string;
  evidence_request_id: string;
  verification_status: string;
  due_at: string;
  is_verified: boolean;
  is_overdue: boolean;
}

export interface EvidenceBundle {
  _id?: string;
  evidence_bundle_id?: string;
  contract_id: string;
  checkpoint: DdayCheckpoint;
  status: "Pending" | "InReview" | "Completed" | "Overdue";
  due_at: string;
  required_count: number;
  submitted_count: number;
  verified_count: number;
  overdue_count: number;
  completion_ratio: number;
  items: EvidenceBundleItem[];
}

export interface EvidenceBundleSummary {
  status: "NotStarted" | "Pending" | "InReview" | "Completed" | "Overdue";
  required_count: number;
  submitted_count: number;
  verified_count: number;
  overdue_count: number;
  completion_ratio: number;
  checkpoints: {
    evidence_bundle_id: string;
    checkpoint: DdayCheckpoint;
    status: string;
    due_at: string;
    completion_ratio: number;
  }[];
}

export interface NotificationRoleSummary {
  sent_count: number;
  read_count: number;
  acknowledged_count: number;
  latest_sent_at: string | null;
}

export interface PreventiveAction {
  _id?: string;
  action_id?: string;
  prevention_case_id: string;
  contract_id: string;
  action_type: string;
  status: PreventiveActionStatus;
  actor_role: string;
  target_role: string;
  requested_at: string;
  due_at: string | null;
  completed_at: string | null;
  note: string | null;
}

/** GET /hug/contracts 목록 행 = 상세 공통 필드. */
export interface HugContractItem {
  contract_id: string;
  property_id: string | null;
  contract_status: string;
  address: Record<string, string>;
  address_summary: string | null;
  guarantee_product: string;
  guarantee_amount: number | null;
  guarantee_status: string;
  deposit: number | null;
  housing_type: string | null;
  contract_start_date: string | null;
  contract_end_date: string | null;
  d_day: number;
  d_day_stage: string | null;
  prediction: AccidentPrediction | null;
  rule_risk: RuleRisk | null;
  prevention_case: PreventionCaseSummary | null;
  prevention_priority: number;
  priority_components: Record<string, number>;
  evidence_bundle: EvidenceBundleSummary;
  notification_status: Record<"tenant" | "landlord" | "hug_admin", NotificationRoleSummary>;
  next_action: string | null;
  owner_center: string | null;
  assignee_user_id: string | null;
}

/** GET /hug/contracts/{id} — 목록 필드 + 상세 전용 필드. */
export interface HugContractDetail extends HugContractItem {
  evidence_bundles: EvidenceBundle[];
  preventive_actions: PreventiveAction[];
  prediction_history: AccidentPrediction[];
  timeline: { event?: string; description?: string; at?: string; created_at?: string; [key: string]: unknown }[];
}

export interface HugContractListData {
  items: HugContractItem[];
  pagination: { page: number; size: number; total_elements: number; total_pages: number };
  as_of_date: string;
  data_mode_filter: "LIVE" | "DEMO";
}

export interface PredictionRefreshBatchResult {
  requested: number;
  succeeded?: number;
  not_scorable?: number;
  failed?: number;
  [key: string]: unknown;
}

export interface PreventionSweepResult {
  checked?: number;
  cases_created?: number;
  actions_created?: number;
  notifications_sent?: number;
  [key: string]: unknown;
}

export interface ContractPreventionData {
  prevention_case?: PreventionCaseSummary | null;
  case?: PreventionCaseSummary | null;
  cases?: PreventionCaseSummary[];
  evidence_bundles?: EvidenceBundle[];
  actions?: PreventiveAction[];
  [key: string]: unknown;
}

File: backend/app/schemas/hug_contract.py
"""HUG 사고 전 계약관리, 사고위험 PoC, 사전예방 API DTO.

사고위험 모델의 출력은 실제 사고확률로 검증된 값이 아니다. ``accident_probability``은
HOUSTA 집계 prior에 평균을 맞춘 미검증 PoC 추정치이며, 응답의 ``calibration_status``와
``basis``를 함께 소비해야 한다.
"""

from __future__ import annotations

from datetime import date
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.provenance import SourceMetadata


PredictionStatus = Literal["SUCCESS", "NOT_SCORABLE", "FAILED"]
PreventionStatus = Literal[
    "Monitoring",
    "RiskDetected",
    "Notified",
    "ActionRequested",
    "EvidenceSubmitted",
    "Verifying",
    "Mitigated",
    "Overdue",
    "EscalatedMonitoring",
]
PreventiveActionStatus = Literal[
    "Requested",
    "InProgress",
    "Submitted",
    "Verifying",
    "Completed",
    "Rejected",
    "Cancelled",
    "Overdue",
]


class AccidentPredictRequest(BaseModel):
    """DB 계약의 권위 있는 입력으로 단건 예측을 생성한다."""

    contract_id: str = Field(min_length=1, max_length=200)


class AccidentPredictionRefreshBatchRequest(BaseModel):
    contract_ids: list[str] | None = Field(default=None, max_length=500)
    data_mode: Literal["LIVE", "DEMO"] = "LIVE"


class AccidentPredictionFactor(BaseModel):
    feature: str
    label: str
    value: str | int | float | None
    importance: float
    explanation_method: Literal["ensemble_global_feature_importance"] = (
        "ensemble_global_feature_importance"
    )


class AccidentPredictionResponse(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    prediction_id: str
    contract_id: str
    pu_risk_score: float | None = None
    risk_percentile: float | None = None
    accident_probability: float | None = None
    calibration_status: str
    prediction_status: PredictionStatus
    failure_reason: list[str] = Field(default_factory=list)
    model_version: str
    model_sha256: str
    feature_snapshot: dict[str, Any]
    top_factors: list[AccidentPredictionFactor] = Field(default_factory=list)
    data_completeness: float = Field(ge=0, le=1)
    basis: str
    predicted_at: str
    valid_until: str
    source: SourceMetadata


class PreventiveActionCreateRequest(BaseModel):
    action_type: Literal[
        "EVIDENCE_REQUEST",
        "CREDIT_ENHANCEMENT_REQUEST",
        "CALLBACK",
        "ASSIGN_OWNER",
        "RERUN_PREDICTION",
        "MANUAL_REVIEW",
    ]
    target_role: Literal["tenant", "landlord", "hug_admin"]
    due_at: str | None = None
    note: str | None = Field(default=None, max_length=2000)
    details: dict[str, Any] = Field(default_factory=dict)


class PreventiveActionUpdateRequest(BaseModel):
    status: PreventiveActionStatus
    note: str | None = Field(default=None, max_length=2000)
    details: dict[str, Any] = Field(default_factory=dict)


class PreventionSweepRequest(BaseModel):
    """`as_of_date`는 고정 시연과 회귀테스트를 위한 결정적 기준일이다."""

    as_of_date: date | None = None
    contract_ids: list[str] | None = Field(default=None, max_length=500)
    data_mode: Literal["LIVE", "DEMO"] = "LIVE"


class EvidenceBundleItemResponse(BaseModel):
    item_key: str
    label: str
    evidence_type: str
    evidence_request_id: str
    verification_status: str
    due_at: str
    is_verified: bool
    is_overdue: bool


class EvidenceBundleResponse(BaseModel):
    evidence_bundle_id: str
    contract_id: str
    checkpoint: Literal["D90", "D60", "D30"]
    policy_version: str
    status: Literal["Pending", "InReview", "Completed", "Overdue"]
    due_at: str
    required_count: int
    submitted_count: int
    verified_count: int
    overdue_count: int
    completion_ratio: float
    items: list[EvidenceBundleItemResponse]
    created_at: str
    updated_at: str


class PreventionCaseResponse(BaseModel):
    prevention_case_id: str
    contract_id: str
    status: PreventionStatus
    triggers: list[dict[str, Any]]
    priority_score: float
    priority_components: dict[str, float]
    owner_user_id: str | None = None
    owner_center: str | None = None
    next_action: str | None = None
    due_at: str | None = None
    policy_version: str
    created_at: str
    updated_at: str


class PreventiveActionResponse(BaseModel):
    action_id: str
    prevention_case_id: str
    contract_id: str
    action_type: str
    status: PreventiveActionStatus
    actor_role: str
    actor_user_id: str | None = None
    target_role: str
    requested_at: str
    due_at: str | None = None
    completed_at: str | None = None
    note: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)
    audit_log: list[dict[str, Any]] = Field(default_factory=list)
